Kernel/Heap/kmalloc.cpp

```diff
@@ -243,11 +317,41 @@ void* kmalloc_eternal(size_t size)
     size = round_up_to_power_of_two(size, sizeof(void*));
 
     SpinlockLocker lock(s_lock);
-    void* ptr = s_next_eternal_ptr;
-    s_next_eternal_ptr += size;
-    VERIFY(s_next_eternal_ptr < s_end_of_eternal_range);
-    g_kmalloc_bytes_eternal += size;
-    return ptr;
+    // Note: Only Eternal heap metadata objects should be in the static eternal heap of the kernel,
+    // therefore we know that each pointer must point to object which is EternalHeap.
+
+    // Note: we iterate over the eternal heaps because we might not allocate their memory ranges
+    // to their full potential capacity. To illustrate what is meant here, consider the following
+    // condition: a caller wants to allocate 10 KiB of eternal memory, so he gets this from
+    // the first eternal heap. Then, a caller wants to allocate 1 MiB of eternal memory,
+    // but because we already allocated 10 KiB of memory in the first eternal heap.
+    // So we need to allocate from a new eternal heap. If we want again to allocate 10 KiB
+    // of memory, we should do that on the first eternal heap because it has plenty
+    // of memory to give and not using it is a waste of precious eternal ranges.
+    // This essentially makes it more expensive to allocate eternal memory in the
+    // time aspect (we iterate on all currently-existing eternal heaps), but comes
+    // with the general idea of not wasting memory as much as possible and minimizing
+    // the usage of eternal heaps as much as possible.
+
+    EternalHeap* eternal_heaps_metadata = reinterpret_cast<EternalHeap*>(kmalloc_eternal_heaps);
+    for (size_t heap_index = 0; heap_index < g_dynamic_eternal_heaps_count; heap_index++) {
+        if (auto* ptr = eternal_heaps_metadata[heap_index].allocate(size); ptr != nullptr)
+            return ptr;
+    }
+
+    // Note: It's very unlikely that the MemoryManager was not initialized at this stage!
+    if (Memory::MemoryManager::is_initialized()) {
+        // Allocate a region. If the allocation fails, assert.
+        auto new_eternal_heap_size = max(size, 1 * MiB);
+        auto region_or_error = MM.allocate_kernel_region(Memory::page_round_up(new_eternal_heap_size), "kmalloc eternal heap", Memory::Region::Access::ReadWrite, AllocationStrategy::AllocateNow);
+        VERIFY(!region_or_error.is_error());
+        new (allocate_new_eternal_heap_slot()) EternalHeap(region_or_error.release_value());
+        g_dynamic_eternal_heaps_count++;
+        if (auto* ptr = eternal_heaps_metadata[g_dynamic_eternal_heaps_count].allocate(size); ptr != nullptr)
```

Comment:
Likewise here:
```suggestion
        auto* ptr = eternal_heaps_metadata[g_dynamic_eternal_heaps_count].allocate(size);
        VERIFY(ptr != nullptr);
        return ptr;
```
Why the `!= nullptr`? Well, [someone](https://github.com/SerenityOS/serenity/pull/10737)'s gonna be upset if the code doesn't do it and it gets merged first.

Also: I think this is an off-by-one. In particular, think about the first time we get here: Before this block (e.g. line 342), we know that `g_dynamic_eternal_heaps_count == 1`, because it was initialized so in `kmalloc_init()`. Then it gets increased to 2 in line 249. Anyway, from the `new …` two lines earlier, we get the pointer to the `EternalHeap` instance. No need to recalculate it. Suggestion:
- Move the `g_dynamic_eternal_heaps_count++;` to `allocate_new_eternal_heap_slot`
- And either use `eternal_heaps_metadata[g_dynamic_eternal_heaps_count - 1]` or the EternalHeap pointer from the `new`.